File: nowcoder/二叉树中和为某一值的路径.py
# encoding: utf-8

'''

@author: liangchi

@contact: bnu_llc@163.com

@software: pycharm

@file: 二叉树中和为某一值的路径.py

@time: 2018/6/14 下午10:03

'''


# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # 返回二维列表，内部每个列表表示找到的路径
    def __init__(self):
        self.res = []
        self.temp = []

    def FindPath(self, root, expectNumber):
        # write code here
        if not root:
            return self.res

        if expectNumber == root.val and not root.left and not root.right:
            self.temp.append(root.val)
            self.res.append(self.temp[:])
            self.temp.pop()

        if root.left:
            self.temp.append(root.val)
            self.FindPath(root.left, expectNumber - root.val)
            self.temp.pop()

        if root.right:
            self.temp.append(root.val)
            self.FindPath(root.right, expectNumber - root.val)
            self.temp.pop()

        return self.res

if __name__ == '__main__':
    pass

File: sword_c++/二叉树中和为某一值的路径.cpp
//
// Created by 李先生 on 2018/8/28.
//

#include <iostream>
#include <string>
#include <vector>
#include <stack>
using namespace std;


struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};

class Solution {
public:
    vector<vector <int>> res;
    vector<int> temp;
    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {

        if (root == NULL)
        {
            return res;
        }
        if (expectNumber == root->val && root->left == NULL && root->right == NULL)
        {
            temp.push_back(root->val);
            res.push_back(temp);
            temp.pop_back();
            return res;
        }
        if (root -> left != NULL)
        {
            temp.push_back(root->val);
            FindPath(root->left, expectNumber - root->val);
            temp.pop_back();
        }
        if (root -> right != NULL)
        {
            temp.push_back(root->val);
            FindPath(root-> right, expectNumber - root->val);
            temp.pop_back();
        }
        return res;
    }
};

int main()
{
    TreeNode * root = new TreeNode(1);
    TreeNode * left_1 = new TreeNode(2);
    TreeNode * right_1 = new TreeNode(3);
    root->left = left_1, root->right = right_1;

    TreeNode * left_2_1 = new TreeNode(4);
    TreeNode * right_2_1 = new TreeNode(5);
    left_1->left = left_2_1, left_1->right = right_2_1;

    TreeNode * left_2_2 = new TreeNode(3);
    TreeNode * right_2_2 = new TreeNode(7);
    right_1->left = left_2_2, right_1->right = right_2_2;

    Solution s;
    vector<vector <int>> res = s.FindPath(root, 7);
    for (int i =0 ; i< res[1].size();i++)
        cout << res[1][i]<<endl;
    return 0;
}